lib/rubygems/safe_marshal/visitors/to_ruby.rb

```diff
@@ -23,39 +24,61 @@ def inspect # :nodoc:
       end
 
       def visit(target)
-        depth = @stack.size
+        stack_idx = @stack_idx
         super
       ensure
-        @stack.slice!(depth.pred..)
+        @stack_idx = stack_idx - 1
       end
 
       private
 
+      def push_stack(element)
+        @stack[@stack_idx] = element
+        @stack_idx += 1
+      end
+
       def visit_Gem_SafeMarshal_Elements_Array(a)
-        register_object([]).replace(a.elements.each_with_index.map do |e, i|
-          @stack << "[#{i}]"
-          visit(e)
-        end)
+        array = register_object([])
+
+        elements = a.elements
+        size = elements.size
+        idx = 0
+        # not idiomatic, but there's a huge number of IMEMOs allocated here, so we avoid the block
+        # because this is such a hot path when doing a bundle install with the full index
+        until idx == size
```

Comment:
Maybe `for idx in (0..size)` maintains the performance while avoiding the bounds check every iteration? Making the range is hopefully cheap.

Reply:
the range unfortunately requires an allocation